### ISSUE TITLE
SET ACCESS METHOD to heap doesn't erase reloptions for partitioned root

### ISSUE THREAD

**Issue by huansong** [greenplum-db/gpdb#15080]
## Bug Report
When change AM of a partitioned table from non-heap to heap, the partition root's reloption isn't changed. For example:
```
postgres=# create table ao2heap(a int, b int) with(appendonly,compresstype=zlib) partition by range(a) (partition p1 start(0) end(1));
CREATE TABLE
postgres=# alter table ao2heap set access method heap;
ALTER TABLE
postgres=# \d+ ao2heap
                            Partitioned table "public.ao2heap"
 Column |  Type   | Collation | Nullable | Default | Storage | Stats target | Description
--------+---------+-----------+----------+---------+---------+--------------+-------------
 a      | integer |           |          |         | plain   |              |
 b      | integer |           |          |         | plain   |              |
Partition key: RANGE (a)
Partitions: ao2heap_1_prt_p1 FOR VALUES FROM (0) TO (1)
Distributed by: (a)
Access method: heap
Options: compresstype=zlib, compresslevel=1
```

The child partition's reloption is correctly erased. But because the root still carries incompatible AM/reloptions, it cannot add partition as is anymore:

```
postgres=# alter table ao2heap add partition p2 start (1) end(2);
ERROR:  unrecognized parameter "compresstype"
```

ao_column->heap is also the same. But it's not a problem for other combinations (heap->ao_row, heap->ao_column, ao_row->ao_column, ao_column->ao_row).

### Greenplum version or build
main branch (HEAD dddc6a950005b446013e9b226b34dd9a027440b4)

**Comment by soumyadeep2007** [comment]
Would your refactor PR https://github.com/greenplum-db/gpdb/pull/14850 somehow fix this @huansong ?

**Comment by huansong** [comment]
@soumyadeep2007 That one won't. But the fix will basically be a refactor too - we need to have better way to deal with ATSETAM with reloptions, which include special handling of partitioned tables that's related to this bug.

**Comment by huansong** [comment]
Fixed here: https://github.com/greenplum-db/gpdb/pull/15101

However, it is not the refactor I promised... Tried some refactor ideas but it turns out to be more messy than I thought. Just note down the refactor idea for future reference:

Ideally, we should have a more clear way to set reloptions when ATSETAM. We should also try to not overload `ATExecSetRelOptions` - currently it carries two additional parameters and has complicated logic when and when not to actually set them. The main idea I had was to have a separate function that gets proposed reloptions w/o setting them, e.g. `ATGetProposedReloptions`. And the logic for ATSETAM in phase 2 becomes more clear:

```
tab->newOptions = ATGetProposedReloptions(&is_storage_option_changed);
if (is_storage_option_changed)
  tab->rewrite |= AT_REWRITE_ALTER_RELOPTS
if (tab->rewrite = 0)
  ATExecSetRelOptions()
```

But eventually I gave up mainly because the new function does too much repetitive work as `ATExecSetRelOptions` and I could only imagine that would bring more maintenance work in future.
